### PR TITLE
Fix #838, fix #842: make GFSM get more data on L4-L7

### LINKED ISSUE #838: GFSM revision for HTTP responses
https://github.com/tempesta-tech/tempesta/pull/837/files#diff-0d15717c2e20453f4271befe0024c2d4R2430 does messy hack to pass request state to GFSM handler. The problem is that GFSM doesn't account requirements of HTTP response handling and this must be fixed.

### LINKED ISSUE #842: Frang: wrong handle of errors during init.
The error is in https://github.com/tempesta-tech/tempesta/blob/e88059d183bb45d63ed62351a83923c0a8038aaa/tempesta_fw/classifier/frang.c#L1118

The function returns non-null `r` on errors. But if error happens during one of `tfw_gfsm_register_hook()` calls, `r` equals to `0`, so the `frang_init()` function returns successful error code (`0`).

### PR DESCRIPTION
**Pull request by krizhanovsky**
1. GFSM #838: use TfwFsmData as L4-L7 data descriptor for GFSM - pass skb and
   off as well as request and response, probably NULL, to each call.
2. GFSM #842: actually hooks are unregisterable for a while;
3. Frang: use explicit uint type conversion instead of FRANG_RESP_TIME_PERIOD
4. Frang: no need to keep per-quantum counter in 64-bit integer;
5. Adjust levels of debug messages on SS layer according to log.h comment;
6. Small cleanups all around.

### CODE REVIEW

File: tempesta_fw/tls.c [hunk 1]
```diff
@@ -109,6 +110,7 @@ tfw_tls_msg_process(void *conn, struct sk_buff *skb, unsigned int off)
 		return TFW_PASS; /* more data needed */
 	} else if (r == 0) {
 		struct sk_buff *nskb;
+		TfwFsmData data = {};
```
Comment (keshonok) [inline tempesta_fw/tls.c]: Just a small note. My understanding is that an empty initializer list is a C++ feature, and also a GNU C extension, but it's not in standard C syntax. There are samples of both `{}` and `{0}` (or `{ 0 }`) throughout Tempesta's code. Perhaps it's best to bring it to one common style.
https://stackoverflow.com/questions/17589533/is-an-empty-initializer-list-valid-c-code/17589839#17589839

File: tempesta_fw/gfsm.c [hunk 2]
```diff
@@ -3,31 +3,63 @@
...
+ * coroutines, i.e. a subroutine can:
+ *   1. call other subroutine;
+ *   2. yield control to other subroutine.
+ *
+ * GFSM is a voluntary (co-operative) data-driven scheduler for the subroutines.
+ * While traditional preemptive (time-sharing) scheduler cares about fair time
+ * slice sharing among processes, GFSM cares about processing hot (in sence of
+ * CPU caches) data by all subroutines, who, and only who, are interested in
+ * the data. I.e. when a packet arrives, all subroutines interested in
+ * processing the packet are called immediatelly while the packet data resides
```
Comment (aleksostapenko) [inline tempesta_fw/gfsm.c]: `immediatelly` -> `immediately`

File: tempesta_fw/gfsm.c [hunk 3]
```diff
@@ -3,31 +3,63 @@
...
  *
+ * A subroutine yields it's control flow by moving to the next state using
+ * tfw_gfsm_move().
+ *
  * Copyright (C) 2014 NatSys Lab. (info@natsys-lab.com).
- * Copyright (C) 2015 Tempesta Technologies, Inc.
+ * Copyright (C) 2015-2018 Tempesta Technologies, Inc.
  *
- * This program is free software; you can redistribute it and/or modify it
+ * This program is free software; you cana redistribute it and/or modify it
```
Comment (aleksostapenko) [inline tempesta_fw/gfsm.c]: `cana` -> `can`

File: tempesta_fw/gfsm.c [hunk 4]
```diff
@@ -125,9 +157,20 @@ tfw_gfsm_switch(TfwGState *st, int state, int prio)
...
 
+/**
+ * TODO #77 (User-kernel space transport): user-space processing is an
+ * asynchronous operation which can be called at different states of different
+ * FSMs, so GFSM must introdufe SLEAL logic:
+ *   1. an FSM (responsible for user-space message mapping) can return
+ *      TFW_STEAL;
+ *   2. getting the return code from gfsm_move() current FSM must finish
+ *      the message processing logic and return with stored current state;
+ *   3. when a user-space program finish GFSM must unwing the call stack and
```
Comment (aleksostapenko) [inline tempesta_fw/gfsm.c]: `unwing` -> `unwind`

File: tempesta_fw/gfsm.c [hunk 5]
```diff
@@ -125,9 +157,20 @@ tfw_gfsm_switch(TfwGState *st, int state, int prio)
 	return fsm_next;
 }
 
+/**
+ * TODO #77 (User-kernel space transport): user-space processing is an
+ * asynchronous operation which can be called at different states of different
+ * FSMs, so GFSM must introdufe SLEAL logic:
```
Comment (aleksostapenko) [inline tempesta_fw/gfsm.c]: `introdufe` -> `introduce` and maybe `SLEAL` -> `STEAL`.

File: tempesta_fw/gfsm.h [hunk 6]
```diff
@@ -129,32 +130,63 @@ enum {
 	TFW_POSTPONE	= SS_POSTPONE,
 };
 
+/**
+ * An input data for a subroutine (FSM) to process.
+ *
+ * Typically an FSM works solely on L4 or L7, however, in futuer, there could be
```
Comment (aleksostapenko) [inline tempesta_fw/gfsm.h]: `futuer` -> `future`.

File: tempesta_fw/classifier/frang.c [hunk 7]
```diff
@@ -541,18 +551,21 @@ frang_http_host_check(const TfwHttpReq *req, FrangAcc *ra)
...
+/**
+ * Monotonically increasing time quantums. The configured time frame is devided
+ * by FRANG_FREQ slots to get the quantums granularity.
+ */
 static unsigned int
 frang_resp_quantum(void)
 {
-	return ((jiffies / HZ) % FRANG_RESP_TIME_PERIOD) * FRANG_FREQ
-		/ frang_cfg.http_resp_code_block->tf;
+	return jiffies / HZ * FRANG_FREQ / frang_cfg.http_resp_code_block->tf;
```
Comment (aleksostapenko) [inline tempesta_fw/classifier/frang.c]: In current implementation if `frang_cfg.http_resp_code_block->tf == 1`, then `ts % FRANG_FREQ` will always give 0.
It seems that `jiffies * FRANG_FREQ / (frang_cfg.http_resp_code_block->tf * HZ)` would produce better granularity.

File: tempesta_fw/http.c [hunk 8]
```diff
@@ -2588,20 +2598,14 @@ tfw_http_resp_cache(TfwHttpMsg *hmresp)
...
-	prev_resp = req->resp;
-	req->resp = hmresp;
-	tfw_gfsm_move(state, TFW_HTTP_FSM_RESP_MSG_FWD, NULL, 0);
-	state->obj = prev_state_obj;
-	req->resp = prev_resp;
+	data.skb = NULL;
+	data.off = 0;
+	data.req = (TfwMsg *)req;
+	data.resp = (TfwMsg *)hmresp;
+	tfw_gfsm_move(&hmresp->conn->state, TFW_HTTP_FSM_RESP_MSG_FWD, &data);
```
Comment (ikoveshnikov) [inline tempesta_fw/http.c]: The hook was added in #717, recently closed #520 required the same for cached responses, according to [Wiki](https://github.com/tempesta-tech/tempesta/wiki/DDoS-mitigation#web-cache-bypass) the feature is required. Sorry but I got confused, is it required to support the #520? Now it's not supported.
If i's required, then either the same gfsm hook must be called in `tfw_http_req_cache_service()` or this hook must be moved to  `tfw_http_resp_fwd()`

Comment (krizhanovsky) [reply tempesta_fw/http.c]: The idea was that if we limit number of negative server response for a client, then it's enough for both of the issues, #520 and #717. The only difference between them is that in #717'th case there is typically the same URI in all client requests while in #520'th URIs are different. There is no need to care about Web cache filling if a client hitted the limit - just let a garbage collector (subject for TDBv2) collect all negative and useless cache entries.

**Pull request merged**